## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
sns.set(color_codes=True)

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score , roc_curve  ,  roc_auc_score

## Overview

In [2]:
train_data = pd.read_csv("/kaggle/input/titanic/train.csv")

In [3]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
test_data = pd.read_csv("/kaggle/input/titanic/test.csv")

In [5]:
test_data.sample(5)

PassengerId  Pclass                                          Name  \
113         1005       3                      Buckley, Miss. Katherine   
39           931       3                                 Hee, Mr. Ling   
405         1297       2  Nourney, Mr. Alfred (Baron von Drachstedt")"   
373         1265       2                        Harbeck, Mr. William H   
289         1181       3                              Ford, Mr. Arthur   

        Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
113  female  18.5      0      0         329944   7.2833   NaN        Q  
39     male   NaN      0      0           1601  56.4958   NaN        S  
405    male  20.0      0      0  SC/PARIS 2166  13.8625   D38        C  
373    male  44.0      0      0         248746  13.0000   NaN        S  
289    male   NaN      0      0       A/5 1478   8.0500   NaN        S

In [6]:
women = train_data.loc[train_data.Sex == 'female']["Survived"]
rate_women = sum(women)/len(women)

print("% of women who survived:", rate_women)

% of women who survived: 0.7420382165605095


In [7]:
y = train_data["Survived"]

features = ["Pclass", "Sex", "SibSp", "Parch"]

In [8]:
men = train_data.loc[train_data.Sex == 'male']["Survived"]
rate_men = sum(men)/len(men)

print("% of men who survived:", rate_men)

% of men who survived: 0.18890814558058924


In [9]:
train_data[features].head()

Pclass     Sex  SibSp  Parch
0       3    male      1      0
1       1  female      1      0
2       3  female      0      0
3       1  female      1      0
4       3    male      0      0

In [10]:
train_data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Handling Outliers

### Droping Cabin col
> Cabin has 687 missing values while total values are 891.

> Missing values are more than 80% of data


In [11]:
train_data= train_data.drop(labels='Cabin', axis=1)

## Mean imputation

In [12]:
train_data['Age'] = train_data['Age'].fillna(train_data['Age'].mean())

## Mode Imputation

In [13]:
train_data['Embarked'] = train_data['Embarked'].fillna(train_data['Embarked'].mode()[0])

In [14]:
train_data.isnull().sum().sum()

0

## One-Hot Encoding

In [15]:
X = pd.get_dummies(train_data[features])
X_test = pd.get_dummies(test_data[features])

In [16]:
X.head()

Pclass  SibSp  Parch  Sex_female  Sex_male
0       3      1      0       False      True
1       1      1      0        True     False
2       3      0      0        True     False
3       1      1      0        True     False
4       3      0      0       False      True

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [18]:
#X_train , X_test , y_train ,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [19]:
y_prob = model.predict_proba(X_test)[:, 1]  # probabilities for the positive class
y_pred = model.predict(X_test)  

In [20]:
roc_auc = roc_auc_score(y_test, y_prob)
print(f'ROC-AUC Score: {(roc_auc*100):.2f}%')


ROC-AUC Score: 86.00%
